usermods/pwm_outputs/usermod_pwm_outputs.h

```diff
@@ -0,0 +1,169 @@
+#pragma once
+#include "wled.h"
+
+#ifndef USERMOD_PWM_OUTPUT_PINS
+  #define USERMOD_PWM_OUTPUT_PINS 3
+#endif
+
+
+class PwmOutput {
+  public:
+
+    PwmOutput() {
+      DEBUG_PRINTLN("pwm_output[-1]: setup disabled output");
+    }
+
+    PwmOutput(int8_t pin, uint32_t freq) : pin_(pin), freq_(freq) {
+      DEBUG_PRINTF("pwm_output[%d]: setup to freq %d\n", pin_, freq_);
+      if (pin_ < 0 || !pinManager.allocatePin(pin_, true, PinOwner::UM_PWM_OUTPUTS)) {
+        return;
+      }
+      
+      channel_ = pinManager.allocateLedc(1);
+      if (channel_ == 255) {
+        DEBUG_PRINTF("pwm_output[%d]: failed to quire ledc\n", pin_);
+        pinManager.deallocatePin(pin_, PinOwner::UM_PWM_OUTPUTS);
+        return;
+      }
+
+      ledcSetup(channel_, freq_, bit_depth_);
+      ledcAttachPin(pin_, channel_);
+      DEBUG_PRINTF("pwm_output[%d]: init successful\n", pin_);
+      enabled_ = true;
+    }
+
+    void close() {
+      DEBUG_PRINTF("pwm_output[%d]: close\n", pin_);
+      if (!enabled_)
+        return;
+      pinManager.deallocatePin(pin_, PinOwner::UM_PWM_OUTPUTS);
+      if (channel_ != 255)
+        pinManager.deallocateLedc(channel_, 1);
+      channel_ = 255;
+      enabled_ = false;
+      duty_ = -1.0f;
+    }
+  
+    void setDuty(const float duty) {
```

Comment:
Perhaps adding an overloaded function that will take `uint8_t` as a parameter would be beneficial so you can pass either 0.0-1.0 range or 0-255 (or 16-bit if you want).

Reply:
Added this `uint8_t` overload, but ended up removing it again, as it was not used. It seemed, that supporting both 0.0-1.0 and 0-255 ranges would not bring much of benefit, but will complicate code and other logic (e.g. what to save in the state json and how to use different ranges when setting duty cycles). Also, for my application, 0-255 range is not suitable because of need for higher resolution.

Reply:
Do not think only about your needs. Think as well what others may need as you are writing a public usermod.
I mentioned `uint8_t` as an example, you can use `uint16_t` with 16-bit resolution and scale it down to available 12-bit.
Someone may tap into your usermod in the future (look into my own PWM_fan usermod how I tapped into Temperature usermod).

Using floats is something I would recommend against if it can be done using integers. Just a personal opinion.

Reply:
Just to clarify, are you suggesting to support multiple ranges (0.0-1.0, 0-255, 0-65535) or just switch to (0-65535)? Switching to 16 bit range instead of float range is easy (and can do if you think it is better), however I am not sure how to support multiple ranges (should it be a configuration per output?). Only supporting 0-255 range is too limiting imo (for me and probably for others), as ESP32 PWM has resolution up to 16bits (future hw might have even more).

In general I am thinking to implement dynamic bit depth (number of steps are automatically selected based on the requested frequency), like in esphome: https://esphome.io/components/output/ledc.html#recommended-frequencies

Reply:
I am only suggesting to implement integer function, if you need 16bit, so be it.

As for PWM resolution: depends on the PWM frequency. You are using a very low frequency which allows you to utilize full resolution. But if I want much higher frequency, resolution would decrease. (get the chart from Espressif)

Reply:
I think I get it now (long wires I guess), added `void setDuty(const uint8_t duty)` and `void setDuty(const uint16_t duty)` overloads (in case some future dependent usermod will need them).

Reply:
Good.
Though my experience with overloading integers like this can sometimes confuse compiler.
To test this add `setDuty(10)` somewhere in the code. You should see compiler complaining about ambiguous overload.
Perhaps a singe 16bit would be enough with info in readme that setDuty() supports float in range [0.0-1.0] and integer in range [0-65535].

Reply:
good point, removed ambiguous overload